### PR TITLE
editor: Unfold buffers with selections on edit + Remove selections on buffer fold

### LINKED ISSUE #36376: Multi-cursor edits in Project Search multi-buffer jump to top after fold toggle
### Summary

Multi-cursor edits in Project Search multi-buffer jump to the top of the first unfolded buffer after toggling a fold

### Description
When editing across multiple results in the Project Search multi-buffer, multi-cursor input is applied to the wrong location after a fold is toggled. Specifically, after collapsing/expanding a fold in the first buffer, subsequent typing or deletion with multiple selections inserts text at the top of the first unfolded buffer instead of at the intended cursor positions.

Steps to reproduce (assuming vim mode is enabled):
1. Open a project with a token that appears in multiple files (e.g., the word foo).
2. Press Cmd+Shift+F (Project Search), search for that token, then open the multi-buffer results.
3. Toggle Excerpt Fold on the first buffer (z a) so the frits buffer gets hidden.
4. Place the caret on the token, and create multiple cursors for that token across the visible buffers by pressing Cmd + Shift + L.
5. Go to insert mode and add characters after the selected token.

**Expected Behavior**:
Every cursor edits its own target position in its respective buffer. No text is inserted elsewhere.
**Actual Behavior**:
The multi-cursor edit inserts characters at the very top of the first unfolded buffer (and/or misapplies edits), instead of doing that only at the selected token positions.

**Demo video**
https://github.com/user-attachments/assets/e508e914-7b9e-41f8-a020-2bc63cea4ef3

### Zed Version and System Specs

Zed: v0.199.9 (Zed) 
OS: macOS 15.6.0
Memory: 48 GiB
Architecture: aarch64

### PR DESCRIPTION
**Pull request by anicioalexandre**
Closes #36376 

Problem:
Multi-cursor edits/selections in multi-buffers view were jumping to incorrect locations after toggling buffer folds. When users created multiple selections across different buffers in a multi-buffer view (like project search results) and then folded one of the buffers, subsequent text insertion would either:

1. Insert text at wrong locations (like at the top of the first unfolded buffer)
2. Replace the entire content in some buffers instead of inserting at the intended cursor positions  
3. Create orphaned selections that caused corruption in the editing experience

The issue seems to happen because when a buffer gets folded in a multi-buffer view, the existing selections associated with that buffer become invalid anchor points.

Solution:
1. Selection Cleanup on Buffer Folding
- Added `remove_selections_from_buffer()` method that filters out all selections from a buffer when it gets folded
- This prevents invalid selections from corrupting subsequent editing operations
- Includes edge case handling: if all selections are removed (all buffers folded), it creates a default selection at the start of the first buffer to prevent panics

2. Unfolding buffers before editing  
- Added `unfold_buffers_with_selections()` call in `handle_input()` ensures buffers with active selections are automatically unfolded before editing
- This helps in fixing an edge case (covered in the tests) where, if you fold all buffers in a multi-buffer view, and try to insert text in a selection, it gets unfolded before the edit happens. Without this, the inserted text would override the entire buffer content.
- If we don't care about this edge case, we could remove this method. I find it ok to add since we already trigger buffer unfolding after edits with `Event::ExcerptsEdited`.

Release Notes:

- Fixed multi-cursor edits jumping to incorrect locations after toggling buffer folds in multi-buffer views (e.g, project search)
  - Multi-cursor selections now properly handle buffer folding/unfolding operations
  - Text insertion no longer occurs at the wrong positions when buffers are folded during multi-cursor editing
  - Eliminated content replacement bugs where entire buffer contents were incorrectly overwritten
  - Added safe fallback behavior when all buffers in a multi-buffer view are folded

### CODE REVIEW

**Comment by anicioalexandre** [comment]
@cla-bot check

**Comment by anicioalexandre** [comment]
@maxbrunsfeld could you have a look when you have some time, please?

File: crates/editor/src/editor.rs [hunk 1]
```diff
@@ -3316,6 +3316,38 @@ impl Editor {
...
+            return;
+        }
+
+        let snapshot = self.buffer.read(cx).snapshot(cx);
+        let buffers_to_unfold: Vec<BufferId> = self
+            .selections
+            .disjoint_anchor_ranges()
+            .filter_map(|range| {
+                // Check if this selection is in a folded buffer
+                if let Some(buffer_id) = snapshot.buffer_id_for_anchor(range.start) {
```
Comment (smitbarmase) [inline crates/editor/src/editor.rs]: Since a selection can span multiple buffers, it makes sense to open all buffers that include that selection, maybe using something like `buffer_ids_for_range`?

File: crates/editor/src/editor.rs [hunk 2]
```diff
@@ -18208,10 +18242,16 @@ impl Editor {
...
             return;
         }
+
         let folded_excerpts = self.buffer().read(cx).excerpts_for_buffer(buffer_id, cx);
         self.display_map.update(cx, |display_map, cx| {
             display_map.fold_buffers([buffer_id], cx)
         });
+
+        self.selections.change_with(cx, |selections| {
+            selections.remove_selections_from_buffer(buffer_id);
```
Comment (smitbarmase) [inline crates/editor/src/editor.rs:18874]: I’m a bit skeptical of this change. I don’t think folding should clear selections inside it. You might fold buffers for various reasons, but in the end it’s just a view and shouldn’t be a destructive action.

That said, I think `editor::SelectAllMatches` shouldn’t select matches in folded buffers. If you are up for making that change.

**Review by smitbarmase — CHANGES_REQUESTED** [review]

File: crates/editor/src/editor.rs (hunk 2 above)
Comment (anicioalexandre) [reply crates/editor/src/editor.rs:18874]: @smitbarmase if we only prevent selecting matches in folded buffers on `editor::SelectAllMatches`, it won't cover the case where users:

1. Select all matches first
2. Then fold a buffer 
3. Then edit multi-buffer selections

By removing selections when folding, we guarantee they won't be there during edits, regardless of operation order. I understand it feels destructive since we won't get the selection back. But both approaches (removing selection when folding vs filtering selections from folded buffers on `editor::SelectAllMatches`) have the same end result.

File: crates/editor/src/editor.rs (hunk 1 above)
Comment (anicioalexandre) [reply crates/editor/src/editor.rs]: nice! I changed to use `buffer_ids_for_range` instead

**Comment by anicioalexandre** [comment]
@smitbarmase do you think we could move forward with this?

File: crates/editor/src/editor.rs [hunk 3]
```diff
@@ -3328,6 +3328,33 @@ impl Editor {
...
+            return;
+        }
+
+        let snapshot = self.buffer.read(cx).snapshot(cx);
+        let buffers_to_unfold: Vec<BufferId> = self
+            .selections
+            .disjoint_anchor_ranges()
+            .flat_map(|range| {
+                // Check all buffers that this selection spans
+                let start_offset = range.start.to_offset(&snapshot);
```
Comment (smitbarmase) [inline crates/editor/src/editor.rs]: I think this conversion might be redundant, since `buffer_ids_for_range` can also take `Range<Anchor>`, right?

File: crates/editor/src/editor.rs [hunk 4]
```diff
@@ -3328,6 +3328,33 @@ impl Editor {
...
+                let end_offset = range.end.to_offset(&snapshot);
+                snapshot.buffer_ids_for_range(start_offset..end_offset)
+            })
+            .filter(|buffer_id| self.is_buffer_folded(*buffer_id, cx))
```
Comment (smitbarmase) [inline crates/editor/src/editor.rs]: I think instead of filtering here, which may lead to multiple duplicate `is_buffer_folded` calls for the same buffer, it might be better to perform this step after creating the HashSet.

File: crates/editor/src/editor.rs [hunk 5]
```diff
@@ -3328,6 +3328,33 @@ impl Editor {
...
+            .collect::<std::collections::HashSet<_>>()
+            .into_iter()
+            .collect();
+
+        if !buffers_to_unfold.is_empty() {
+            self.display_map.update(cx, |display_map, cx| {
+                display_map.unfold_buffers(buffers_to_unfold.iter().copied(), cx);
+            });
```
Comment (smitbarmase) [inline crates/editor/src/editor.rs]: I think we should also emit the `EditorEvent::BufferFoldToggled` event here, since it affects other parts of Zed.

**Review by smitbarmase — CHANGES_REQUESTED** [review]
I added few more comments, I think we can merge after those are resolved.

**Comment by smitbarmase** [comment]
Could you also rebase this on the latest main?

**Review by smitbarmase — APPROVED** [review]

**Comment by smitbarmase** [comment]
Thank you!
